src/test/atlas_planned_maintenance_testing/mod.rs

```diff
@@ -0,0 +1,103 @@
+mod json_models;
+
+use std::{env, fs::File, io::BufWriter, panic::AssertUnwindSafe};
+
+use futures::FutureExt;
+use serde_json::{from_str, from_value, Value};
+use time::OffsetDateTime;
+
+use crate::{
+    bson::{doc, Bson},
+    test::spec::unified_runner::{entity::Entity, test_file::TestFile, test_runner::TestRunner},
+};
+
+use json_models::{Events, Results};
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn workload_executor() {
+    if env::var("ATLAS_PLANNED_MAINTENANCE_TESTING").is_err() {
+        // This test should only be run from the workload-executor script.
+        return;
+    }
+
+    let connection_string =
+        env::var("WORKLOAD_EXECUTOR_CONNECTION_STRING").expect("No connection string specified");
+
+    let workload_string = env::var("WORKLOAD_EXECUTOR_WORKLOAD").expect("No workload specified");
+    let workload = from_str(&workload_string).expect("Error converting workload to JSON");
+
+    let mut test_runner = TestRunner::new_with_connection_string(&connection_string).await;
+
+    let execution_errors = execute_workload(&mut test_runner, workload).await;
+    let events = populate_and_write_events(&mut test_runner, execution_errors);
+    write_results(&mut test_runner, events);
+}
+
+async fn execute_workload(test_runner: &mut TestRunner, workload: Value) -> Vec<Bson> {
+    let mut execution_errors: Vec<Bson> = vec![];
+
+    let test_file: TestFile = from_value(workload).unwrap();
+    let description = test_file.description.clone();
+    if AssertUnwindSafe(test_runner.run_test(test_file, |_| true))
+        .catch_unwind()
+        .await
+        .is_err()
+    {
+        execution_errors.push(
+            doc! {
+                "error": format!("Unexpected error occurred while running {}", description),
```

Comment:
is the problem that `run_test` uses `panic` to report failures? the Swift runner has a couple custom error types we throw for unexpected failures, I wonder how much work it would be to switch to something like that so `run_test` could return a `Result`? otherwise I imagine this is going to make it really difficult to debug any failures we see in Evergreen

Reply:
thanks for the context! 

in Swift, the top-level method `UnifiedTestRunner.runTests` is marked `throws`, so we can propagate errors all the way up to that level pretty easily, which leads the calling test case to throw an error and fail. 
for operations that expect errors we use a `do`/`catch` around the appropriate code, and for some unexpected errors we catch them and re-throw them with extra context added. some unexpected errors just get directly propagated up though, like anything during test setup I believe.

I suppose this all ends up being more straightforward and concise in Swift than Rust since the Swift compiler does not care about what type of error a method throws, so we don't have to worry about converting errors all to some consistent type. also, we can just `try` things that throw in Swift, but since we can't use `?` on async return values working with the results is a lot more verbose here. 🤔

Reply:
I think it's reasonable to take this approach here; it should work for what's needed and it keeps this PR from ballooning.

I do think there should be a bug to track converting the test code to return errors rather than using `panic!`.  I'm not sure how unwind-safe the test runner actually is; I don't think there are going to be memory-safety problems, but it seems quite possible the runner could end up with broken assumptions that cause flaky tests.  It seems roughly equivalent to the cancellation issue.

Reply:
I agree with Abraham, deferring test runner refactoring to use errors to a follow up ticket SGTM